{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NBL Analysis 1 - Identifying Risk-Associated Mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Neuroblastoma (NBL) is a cancer that develops from immature nerve cells of the sympathetic nervous system, primarily affecting infants and children. In this analysis, I will be analyzing the genomic data of NBL samples from the Therapeutically Applicable Research to Generate Effective Treatments (TARGET) project to identify predictive mutation/gene expression features that could predict risk of patients, potentially informing the treatment strategy.\n",
    "\n",
    "First, I will need to import the TARGET NBL somatic mutation data in MAF format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "cwd = os.getcwd()\n",
    "DATA = cwd + '/data'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>...</th>\n",
       "      <th>CGC_Tumor_Types_Germline</th>\n",
       "      <th>CGC_Other_Diseases</th>\n",
       "      <th>DNARepairGenes_Role</th>\n",
       "      <th>FamilialCancerDatabase_Syndromes</th>\n",
       "      <th>MUTSIG_Published_Results</th>\n",
       "      <th>OREGANNO_ID</th>\n",
       "      <th>OREGANNO_Values</th>\n",
       "      <th>PolyPhen</th>\n",
       "      <th>SIFT</th>\n",
       "      <th>RNAseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL2</td>\n",
       "      <td>27</td>\n",
       "      <td>broadinstitute.org</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>179077416</td>\n",
       "      <td>179077416</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>possibly_damaging(0.609)</td>\n",
       "      <td>tolerated(0.24)</td>\n",
       "      <td>RNAseq_support</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ALK</td>\n",
       "      <td>238</td>\n",
       "      <td>broadinstitute.org</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>29445213</td>\n",
       "      <td>29445213</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neuroblastoma_Familial_Clustering_of|Congenita...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>probably_damaging(0.99)</td>\n",
       "      <td>deleterious(0)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ALK</td>\n",
       "      <td>238</td>\n",
       "      <td>broadinstitute.org</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>29443697</td>\n",
       "      <td>29443697</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neuroblastoma_Familial_Clustering_of|Congenita...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>possibly_damaging(0.748)</td>\n",
       "      <td>deleterious(0)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALK</td>\n",
       "      <td>238</td>\n",
       "      <td>broadinstitute.org</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>29443697</td>\n",
       "      <td>29443697</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neuroblastoma_Familial_Clustering_of|Congenita...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>benign(0.082)</td>\n",
       "      <td>deleterious(0.02)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALK</td>\n",
       "      <td>238</td>\n",
       "      <td>broadinstitute.org</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>29443696</td>\n",
       "      <td>29443696</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neuroblastoma_Familial_Clustering_of|Congenita...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>probably_damaging(0.993)</td>\n",
       "      <td>deleterious(0)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 77 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id              Center  NCBI_Build Chromosome  \\\n",
       "0        ABL2              27  broadinstitute.org          37          1   \n",
       "1         ALK             238  broadinstitute.org          37          2   \n",
       "2         ALK             238  broadinstitute.org          37          2   \n",
       "3         ALK             238  broadinstitute.org          37          2   \n",
       "4         ALK             238  broadinstitute.org          37          2   \n",
       "\n",
       "   Start_position  End_position Strand Variant_Classification Variant_Type  \\\n",
       "0       179077416     179077416      +      Missense_Mutation          SNP   \n",
       "1        29445213      29445213      +      Missense_Mutation          SNP   \n",
       "2        29443697      29443697      +      Missense_Mutation          SNP   \n",
       "3        29443697      29443697      +      Missense_Mutation          SNP   \n",
       "4        29443696      29443696      +      Missense_Mutation          SNP   \n",
       "\n",
       "   ... CGC_Tumor_Types_Germline CGC_Other_Diseases DNARepairGenes_Role  \\\n",
       "0  ...                      NaN                NaN                 NaN   \n",
       "1  ...            neuroblastoma                NaN                 NaN   \n",
       "2  ...            neuroblastoma                NaN                 NaN   \n",
       "3  ...            neuroblastoma                NaN                 NaN   \n",
       "4  ...            neuroblastoma                NaN                 NaN   \n",
       "\n",
       "                    FamilialCancerDatabase_Syndromes  \\\n",
       "0                                                NaN   \n",
       "1  Neuroblastoma_Familial_Clustering_of|Congenita...   \n",
       "2  Neuroblastoma_Familial_Clustering_of|Congenita...   \n",
       "3  Neuroblastoma_Familial_Clustering_of|Congenita...   \n",
       "4  Neuroblastoma_Familial_Clustering_of|Congenita...   \n",
       "\n",
       "   MUTSIG_Published_Results OREGANNO_ID OREGANNO_Values  \\\n",
       "0                       NaN         NaN             NaN   \n",
       "1                       NaN         NaN             NaN   \n",
       "2                       NaN         NaN             NaN   \n",
       "3                       NaN         NaN             NaN   \n",
       "4                       NaN         NaN             NaN   \n",
       "\n",
       "                   PolyPhen               SIFT          RNAseq  \n",
       "0  possibly_damaging(0.609)    tolerated(0.24)  RNAseq_support  \n",
       "1   probably_damaging(0.99)     deleterious(0)             NaN  \n",
       "2  possibly_damaging(0.748)     deleterious(0)             NaN  \n",
       "3             benign(0.082)  deleterious(0.02)             NaN  \n",
       "4  probably_damaging(0.993)     deleterious(0)             NaN  \n",
       "\n",
       "[5 rows x 77 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Reading in the maf file\n",
    "maf_file = \"/TARGET_NBL_WXS_somatic_verified.maf.txt\"\n",
    "maf = pd.read_csv(DATA + maf_file, comment='#', sep=\"\\t\")\n",
    "maf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To identify which mutations are risk-associated with NBL, we can first look at which gene is the most mutated gene. This can be found by counting how many patients have the mutation for each gene. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hugo_Symbol\n",
       "ALK         19\n",
       "PTPN11       6\n",
       "ATRX         6\n",
       "SPTA1        6\n",
       "PMFBP1       4\n",
       "            ..\n",
       "PDCD1LG2     1\n",
       "PDGFA        1\n",
       "PDGFRA       1\n",
       "PIK3C2G      1\n",
       "ABL2         1\n",
       "Name: Case_USI, Length: 114, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Use groupby to count how many patients have the mutation for each gene\n",
    "most_mutated_gene = maf.groupby(\"Hugo_Symbol\").Case_USI.size()\n",
    "most_mutated_gene.sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It seems like 19 patients have mutations in the ALK gene, which makes it the most mutated gene in the dataset. Now, we can investigate which mutation is most prevalent. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Genome_Change\n",
       "g.chr2:29432664C>T       7\n",
       "g.chr2:29443695G>T       5\n",
       "g.chr2:16082317C>T       3\n",
       "g.chr12:112888198G>A     2\n",
       "g.chr2:29443696A>C       2\n",
       "                        ..\n",
       "g.chr3:130289800G>T      1\n",
       "g.chr3:12633211G>C       1\n",
       "g.chr3:118623540C>G      1\n",
       "g.chr3:10084803C>A       1\n",
       "g.chr10:104162112delC    1\n",
       "Name: Case_USI, Length: 163, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Use groupby to count the occurance of a unique mutation in the patients\n",
    "most_prevalent_mutation = maf.groupby(\"Genome_Change\").Case_USI.size()\n",
    "most_prevalent_mutation.sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The most prevalent mutation is the mutation of C to T in chromosome 2, location 29432664, which had a total of 7 patients having this particular mutation.\n",
    "\n",
    "Next, we can look into other cancer datasets to observe if the putations in pediatric NBL are also observed. The cancer dataset we will be using here is [Cancer Hotspots](https://www.cancerhotspots.org/#/home) database. It is a database collecting frequently observed mutations from public cancer datasets. We will check how many mutations in the TARGET dataset also appear in Cancer Hotspots dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele1</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>HGVSp_Short</th>\n",
       "      <th>TUMORTYPE</th>\n",
       "      <th>PLATFORM</th>\n",
       "      <th>judgement</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WARS2</td>\n",
       "      <td>10352</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>119575617</td>\n",
       "      <td>119575617</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1000G&gt;A</td>\n",
       "      <td>p.Val334Ile</td>\n",
       "      <td>p.V334I</td>\n",
       "      <td>acyc</td>\n",
       "      <td>exome</td>\n",
       "      <td>RETAIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OPN3</td>\n",
       "      <td>23596</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>241761094</td>\n",
       "      <td>241761094</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>c.899C&gt;T</td>\n",
       "      <td>p.Ser300Leu</td>\n",
       "      <td>p.S300L</td>\n",
       "      <td>acyc</td>\n",
       "      <td>exome</td>\n",
       "      <td>RETAIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NAA16</td>\n",
       "      <td>79612</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>13</td>\n",
       "      <td>41894863</td>\n",
       "      <td>41894863</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>c.305G&gt;A</td>\n",
       "      <td>p.Cys102Tyr</td>\n",
       "      <td>p.C102Y</td>\n",
       "      <td>acyc</td>\n",
       "      <td>exome</td>\n",
       "      <td>RETAIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DHODH</td>\n",
       "      <td>1723</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>16</td>\n",
       "      <td>72056354</td>\n",
       "      <td>72056354</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>c.799A&gt;C</td>\n",
       "      <td>p.Ile267Leu</td>\n",
       "      <td>p.I267L</td>\n",
       "      <td>acyc</td>\n",
       "      <td>exome</td>\n",
       "      <td>RETAIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KRTAP21-1</td>\n",
       "      <td>337977</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>21</td>\n",
       "      <td>32127473</td>\n",
       "      <td>32127473</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.224G&gt;A</td>\n",
       "      <td>p.Cys75Tyr</td>\n",
       "      <td>p.C75Y</td>\n",
       "      <td>acyc</td>\n",
       "      <td>exome</td>\n",
       "      <td>RETAIN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  Start_Position  \\\n",
       "0       WARS2           10352      .     GRCh37          1       119575617   \n",
       "1        OPN3           23596      .     GRCh37          1       241761094   \n",
       "2       NAA16           79612      .     GRCh37         13        41894863   \n",
       "3       DHODH            1723      .     GRCh37         16        72056354   \n",
       "4   KRTAP21-1          337977      .     GRCh37         21        32127473   \n",
       "\n",
       "   End_Position Strand Variant_Classification Variant_Type Reference_Allele  \\\n",
       "0     119575617      +      Missense_Mutation          SNP                C   \n",
       "1     241761094      +      Missense_Mutation          SNP                G   \n",
       "2      41894863      +      Missense_Mutation          SNP                G   \n",
       "3      72056354      +      Missense_Mutation          SNP                A   \n",
       "4      32127473      +      Missense_Mutation          SNP                C   \n",
       "\n",
       "  Tumor_Seq_Allele1 Tumor_Seq_Allele2      HGVSc        HGVSp HGVSp_Short  \\\n",
       "0                 C                 T  c.1000G>A  p.Val334Ile     p.V334I   \n",
       "1                 G                 A   c.899C>T  p.Ser300Leu     p.S300L   \n",
       "2                 G                 A   c.305G>A  p.Cys102Tyr     p.C102Y   \n",
       "3                 A                 C   c.799A>C  p.Ile267Leu     p.I267L   \n",
       "4                 C                 T   c.224G>A   p.Cys75Tyr      p.C75Y   \n",
       "\n",
       "  TUMORTYPE PLATFORM judgement  \n",
       "0      acyc    exome    RETAIN  \n",
       "1      acyc    exome    RETAIN  \n",
       "2      acyc    exome    RETAIN  \n",
       "3      acyc    exome    RETAIN  \n",
       "4      acyc    exome    RETAIN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hotspot_file = \"/cancerhotspots.v2.light.maf.gz\"\n",
    "hotspot = pd.read_csv(DATA + hotspot_file, comment='#', sep='\\t')\n",
    "hotspot.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To easily do this, I created two dataframes, each for TARGET dataset and hotspot dataset, performed an inner join, and dropped duplicates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Tumor_Seq_Allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>179077416</td>\n",
       "      <td>179077416</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>29445213</td>\n",
       "      <td>29445213</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>29443697</td>\n",
       "      <td>29443697</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>29443696</td>\n",
       "      <td>29443696</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>29443695</td>\n",
       "      <td>29443695</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>17</td>\n",
       "      <td>7577097</td>\n",
       "      <td>7577097</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>19</td>\n",
       "      <td>12384448</td>\n",
       "      <td>12384448</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>19</td>\n",
       "      <td>37038544</td>\n",
       "      <td>37038544</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>19</td>\n",
       "      <td>37037970</td>\n",
       "      <td>37037970</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>19</td>\n",
       "      <td>40542140</td>\n",
       "      <td>40542140</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>130 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromosome  Start_Position  End_Position Tumor_Seq_Allele\n",
       "0            1       179077416     179077416                C\n",
       "1            2        29445213      29445213                T\n",
       "2            2        29443697      29443697                G\n",
       "4            2        29443696      29443696                C\n",
       "8            2        29443695      29443695                T\n",
       "..         ...             ...           ...              ...\n",
       "422         17         7577097       7577097                T\n",
       "436         19        12384448      12384448                -\n",
       "447         19        37038544      37038544                A\n",
       "448         19        37037970      37037970                C\n",
       "449         19        40542140      40542140                A\n",
       "\n",
       "[130 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create two dataframes that contain chromosome, position, and allele info and join the sets\n",
    "target_mut = pd.DataFrame().assign(Chromosome=maf['Chromosome'], Start_Position=maf['Start_position'], End_Position=maf['End_position'], Tumor_Seq_Allele=maf['Tumor_Seq_Allele1'])\n",
    "hotspot_mut = pd.DataFrame().assign(Chromosome=hotspot['Chromosome'], Start_Position=hotspot['Start_Position'], End_Position=hotspot['End_Position'], Tumor_Seq_Allele=hotspot['Tumor_Seq_Allele2'])\n",
    "\n",
    "# Do inner join and drop duplicates to get unique mutations\n",
    "target_mut.merge(hotspot_mut, how = 'inner' ,indicator=False).drop_duplicates()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The resulting dataframe has 130 rows, meaning that there are 130 unique mutations in the TARGET NBL dataset also found in Cancer Hotspot dataset. \n",
    "\n",
    "Now, we can explore the mutation information and clinical information at the same time. The clinical information for the TARGET dataset has been prepared and provided for the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TARGET USI</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Age at Diagnosis in Days</th>\n",
       "      <th>First Event</th>\n",
       "      <th>Event Free Survival Time in Days</th>\n",
       "      <th>Vital Status</th>\n",
       "      <th>Overall Survival Time in Days</th>\n",
       "      <th>Year of Diagnosis</th>\n",
       "      <th>...</th>\n",
       "      <th>Sites of Disease Involvement</th>\n",
       "      <th>Site of Relapse</th>\n",
       "      <th>Relapse Percent Tumor</th>\n",
       "      <th>Relapse Percent Necrosis</th>\n",
       "      <th>Relapse Percent Tumor v/s Stroma</th>\n",
       "      <th>Comment</th>\n",
       "      <th>5yr_efs</th>\n",
       "      <th>in_rnaseq</th>\n",
       "      <th>in_maf</th>\n",
       "      <th>train_test</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TARGET-30-PAHYWC</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>704</td>\n",
       "      <td>Event</td>\n",
       "      <td>324.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>437.0</td>\n",
       "      <td>1993</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high risk</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TARGET-30-PAICGF</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1278</td>\n",
       "      <td>Event</td>\n",
       "      <td>772.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1314.0</td>\n",
       "      <td>1994</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high risk</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TARGET-30-PAIFXV</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>2004</td>\n",
       "      <td>Event</td>\n",
       "      <td>630.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1114.0</td>\n",
       "      <td>1995</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high risk</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TARGET-30-PAILNU</td>\n",
       "      <td>Male</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1683</td>\n",
       "      <td>Event</td>\n",
       "      <td>647.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>761.0</td>\n",
       "      <td>1997</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high risk</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TARGET-30-PAIMDT</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1408</td>\n",
       "      <td>Death</td>\n",
       "      <td>878.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>878.0</td>\n",
       "      <td>1997</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high risk</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         TARGET USI  Gender     Race Ethnicity  Age at Diagnosis in Days  \\\n",
       "0  TARGET-30-PAHYWC    Male    White   Unknown                       704   \n",
       "1  TARGET-30-PAICGF  Female    White   Unknown                      1278   \n",
       "2  TARGET-30-PAIFXV  Female    White   Unknown                      2004   \n",
       "3  TARGET-30-PAILNU    Male  Unknown   Unknown                      1683   \n",
       "4  TARGET-30-PAIMDT  Female    White   Unknown                      1408   \n",
       "\n",
       "  First Event  Event Free Survival Time in Days Vital Status  \\\n",
       "0       Event                             324.0         Dead   \n",
       "1       Event                             772.0         Dead   \n",
       "2       Event                             630.0         Dead   \n",
       "3       Event                             647.0         Dead   \n",
       "4       Death                             878.0         Dead   \n",
       "\n",
       "   Overall Survival Time in Days  Year of Diagnosis  ...  \\\n",
       "0                          437.0               1993  ...   \n",
       "1                         1314.0               1994  ...   \n",
       "2                         1114.0               1995  ...   \n",
       "3                          761.0               1997  ...   \n",
       "4                          878.0               1997  ...   \n",
       "\n",
       "   Sites of Disease Involvement Site of Relapse Relapse Percent Tumor  \\\n",
       "0                           NaN             NaN                   NaN   \n",
       "1                           NaN             NaN                   NaN   \n",
       "2                           NaN             NaN                   NaN   \n",
       "3                           NaN             NaN                   NaN   \n",
       "4                           NaN             NaN                   NaN   \n",
       "\n",
       "  Relapse Percent Necrosis Relapse Percent Tumor v/s Stroma  Comment  \\\n",
       "0                      NaN                              NaN      NaN   \n",
       "1                      NaN                              NaN      NaN   \n",
       "2                      NaN                              NaN      NaN   \n",
       "3                      NaN                              NaN      NaN   \n",
       "4                      NaN                              NaN      NaN   \n",
       "\n",
       "     5yr_efs in_rnaseq in_maf train_test  \n",
       "0  high risk     False   True      train  \n",
       "1  high risk     False   True      train  \n",
       "2  high risk      True  False      train  \n",
       "3  high risk     False   True       test  \n",
       "4  high risk     False   True      train  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_file = \"/nbl_target_clinical.txt\"\n",
    "clinical = pd.read_csv(DATA + clinical_file, comment='#', sep=\"\\t\")\n",
    "clinical.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `TARGET USI` column indicates the patient ID, `5yr_efs` column indicates whether the patient belongs to the 'high risk' group (progression, relapse, or death within the 5 years of the genomic profiling) or the 'low risk' group (no event for more than 5 years). The `train_test` column indicates whether the patient belongs to training or test set for the machine learning model that we will build in the later part of the analysis. We will focus on  the `5yr_efs` information and investigate if the most mutated gene has association with high risk of disease development events."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(202, 36)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(103,)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maf['Case_USI'].unique().shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The clinical dataset has 202 rows, while there are 103 unique patients in the original TARGET dataset. Therefore, we have to subset the clinical dataset to only include the 103 unique patients from TARGET dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(103, 37)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a list of unique Case_USI in maf file\n",
    "unique_usi = maf['Case_USI'].tolist()\n",
    "\n",
    "# Subset the clinical table to containi only the 103 unique patients from TARGET\n",
    "clinical_103 = clinical[clinical['TARGET USI'].isin(unique_usi)]\n",
    "clinical_103 = clinical_103.reset_index()\n",
    "clinical_103.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From previous steps, we have identified that the most mutated gene is the 'ALK' gene. We will create an additional column in the clinical dataframe to keep track of which patient has a mutation in 'ALK'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TARGET-30-PAKXDZ',\n",
       " 'TARGET-30-PANRVJ',\n",
       " 'TARGET-30-PARKNP',\n",
       " 'TARGET-30-PARLTG',\n",
       " 'TARGET-30-PARZIP',\n",
       " 'TARGET-30-PAPRPR',\n",
       " 'TARGET-30-PASVRU',\n",
       " 'TARGET-30-PATFXV',\n",
       " 'TARGET-30-PALAKE',\n",
       " 'TARGET-30-PANZVU',\n",
       " 'TARGET-30-PATFCY',\n",
       " 'TARGET-30-PAINLN',\n",
       " 'TARGET-30-PANWRR',\n",
       " 'TARGET-30-PANXJL',\n",
       " 'TARGET-30-PAPTFZ',\n",
       " 'TARGET-30-PAPTLV',\n",
       " 'TARGET-30-PARURB',\n",
       " 'TARGET-30-PALNLU',\n",
       " 'TARGET-30-PANBCI']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract the rows from maf so that we get a dataframe that shows which patient ids have a mutation in ALK\n",
    "target_alk = pd.DataFrame().assign(patient_id=maf['Case_USI'], Hugo_Symbol=maf['Hugo_Symbol'])\n",
    "target_alk = target_alk[1:20]\n",
    "\n",
    "# Convert it to a list\n",
    "usi_list = target_alk['patient_id'].tolist()\n",
    "usi_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-19-5784f0ac93a3>:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinical_103['TMG'][i] = True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>TARGET USI</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Age at Diagnosis in Days</th>\n",
       "      <th>First Event</th>\n",
       "      <th>Event Free Survival Time in Days</th>\n",
       "      <th>Vital Status</th>\n",
       "      <th>Overall Survival Time in Days</th>\n",
       "      <th>...</th>\n",
       "      <th>Site of Relapse</th>\n",
       "      <th>Relapse Percent Tumor</th>\n",
       "      <th>Relapse Percent Necrosis</th>\n",
       "      <th>Relapse Percent Tumor v/s Stroma</th>\n",
       "      <th>Comment</th>\n",
       "      <th>5yr_efs</th>\n",
       "      <th>in_rnaseq</th>\n",
       "      <th>in_maf</th>\n",
       "      <th>train_test</th>\n",
       "      <th>TMG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TARGET-30-PAHYWC</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>704</td>\n",
       "      <td>Event</td>\n",
       "      <td>324.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>437.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high risk</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TARGET-30-PAICGF</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1278</td>\n",
       "      <td>Event</td>\n",
       "      <td>772.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1314.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high risk</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>TARGET-30-PAILNU</td>\n",
       "      <td>Male</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1683</td>\n",
       "      <td>Event</td>\n",
       "      <td>647.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>761.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high risk</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>test</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>TARGET-30-PAIMDT</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1408</td>\n",
       "      <td>Death</td>\n",
       "      <td>878.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>878.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high risk</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>TARGET-30-PAINLN</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>1404</td>\n",
       "      <td>Death</td>\n",
       "      <td>599.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>599.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high risk</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>190</td>\n",
       "      <td>TARGET-30-PATFCY</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>1040</td>\n",
       "      <td>Death</td>\n",
       "      <td>787.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>787.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high risk</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>191</td>\n",
       "      <td>TARGET-30-PATFIN</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>631</td>\n",
       "      <td>Event</td>\n",
       "      <td>1192.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1363.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high risk</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>192</td>\n",
       "      <td>TARGET-30-PATFXV</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Hispanic or Latino</td>\n",
       "      <td>648</td>\n",
       "      <td>Progression</td>\n",
       "      <td>236.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>296.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Primary site; Bone; Other metastatic sites</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high risk</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>195</td>\n",
       "      <td>TARGET-30-PATGWT</td>\n",
       "      <td>Male</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>591</td>\n",
       "      <td>Relapse</td>\n",
       "      <td>727.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>928.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Bone; Lymph Nodes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high risk</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>197</td>\n",
       "      <td>TARGET-30-PATHVK</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>2109</td>\n",
       "      <td>Progression</td>\n",
       "      <td>171.0</td>\n",
       "      <td>Alive</td>\n",
       "      <td>2866.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high risk</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>103 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     index        TARGET USI  Gender                       Race  \\\n",
       "0        0  TARGET-30-PAHYWC    Male                      White   \n",
       "1        1  TARGET-30-PAICGF  Female                      White   \n",
       "2        3  TARGET-30-PAILNU    Male                    Unknown   \n",
       "3        4  TARGET-30-PAIMDT  Female                      White   \n",
       "4        5  TARGET-30-PAINLN    Male                      White   \n",
       "..     ...               ...     ...                        ...   \n",
       "98     190  TARGET-30-PATFCY    Male                      White   \n",
       "99     191  TARGET-30-PATFIN  Female                      White   \n",
       "100    192  TARGET-30-PATFXV  Female                      White   \n",
       "101    195  TARGET-30-PATGWT    Male  Black or African American   \n",
       "102    197  TARGET-30-PATHVK  Female                      White   \n",
       "\n",
       "                  Ethnicity  Age at Diagnosis in Days  First Event  \\\n",
       "0                   Unknown                       704        Event   \n",
       "1                   Unknown                      1278        Event   \n",
       "2                   Unknown                      1683        Event   \n",
       "3                   Unknown                      1408        Death   \n",
       "4                   Unknown                      1404        Death   \n",
       "..                      ...                       ...          ...   \n",
       "98   Not Hispanic or Latino                      1040        Death   \n",
       "99   Not Hispanic or Latino                       631        Event   \n",
       "100      Hispanic or Latino                       648  Progression   \n",
       "101  Not Hispanic or Latino                       591      Relapse   \n",
       "102  Not Hispanic or Latino                      2109  Progression   \n",
       "\n",
       "     Event Free Survival Time in Days Vital Status  \\\n",
       "0                               324.0         Dead   \n",
       "1                               772.0         Dead   \n",
       "2                               647.0         Dead   \n",
       "3                               878.0         Dead   \n",
       "4                               599.0         Dead   \n",
       "..                                ...          ...   \n",
       "98                              787.0         Dead   \n",
       "99                             1192.0         Dead   \n",
       "100                             236.0         Dead   \n",
       "101                             727.0         Dead   \n",
       "102                             171.0        Alive   \n",
       "\n",
       "     Overall Survival Time in Days  ...  \\\n",
       "0                            437.0  ...   \n",
       "1                           1314.0  ...   \n",
       "2                            761.0  ...   \n",
       "3                            878.0  ...   \n",
       "4                            599.0  ...   \n",
       "..                             ...  ...   \n",
       "98                           787.0  ...   \n",
       "99                          1363.0  ...   \n",
       "100                          296.0  ...   \n",
       "101                          928.0  ...   \n",
       "102                         2866.0  ...   \n",
       "\n",
       "                                Site of Relapse  Relapse Percent Tumor  \\\n",
       "0                                           NaN                    NaN   \n",
       "1                                           NaN                    NaN   \n",
       "2                                           NaN                    NaN   \n",
       "3                                           NaN                    NaN   \n",
       "4                                           NaN                    NaN   \n",
       "..                                          ...                    ...   \n",
       "98                                          NaN                    NaN   \n",
       "99                                          NaN                    NaN   \n",
       "100  Primary site; Bone; Other metastatic sites                    NaN   \n",
       "101                           Bone; Lymph Nodes                    NaN   \n",
       "102                                 Bone Marrow                    NaN   \n",
       "\n",
       "    Relapse Percent Necrosis Relapse Percent Tumor v/s Stroma Comment  \\\n",
       "0                        NaN                              NaN     NaN   \n",
       "1                        NaN                              NaN     NaN   \n",
       "2                        NaN                              NaN     NaN   \n",
       "3                        NaN                              NaN     NaN   \n",
       "4                        NaN                              NaN     NaN   \n",
       "..                       ...                              ...     ...   \n",
       "98                       NaN                              NaN     NaN   \n",
       "99                       NaN                              NaN     NaN   \n",
       "100                      NaN                              NaN     NaN   \n",
       "101                      NaN                              NaN     NaN   \n",
       "102                      NaN                              NaN     NaN   \n",
       "\n",
       "       5yr_efs  in_rnaseq in_maf train_test    TMG  \n",
       "0    high risk      False   True      train  False  \n",
       "1    high risk      False   True      train  False  \n",
       "2    high risk      False   True       test  False  \n",
       "3    high risk      False   True      train  False  \n",
       "4    high risk      False   True      train   True  \n",
       "..         ...        ...    ...        ...    ...  \n",
       "98   high risk       True   True      train   True  \n",
       "99   high risk      False   True      train  False  \n",
       "100  high risk       True   True      train   True  \n",
       "101  high risk      False   True      train  False  \n",
       "102  high risk       True   True      train  False  \n",
       "\n",
       "[103 rows x 38 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initialize a new column and fill it up with a default boolean for now\n",
    "clinical_103['TMG'] = False\n",
    "\n",
    "# Chang the value of TMG column if the given usi has a mutation in ALK\n",
    "for i, usi_103 in enumerate(clinical_103['TARGET USI']):\n",
    "    if usi_103 in usi_list:\n",
    "        clinical_103['TMG'][i] = True\n",
    "    \n",
    "clinical_103.head(103)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can create a count table to easily visualize the number of patients satisfying each condition."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>5yr_efs</th>\n",
       "      <th>high risk</th>\n",
       "      <th>low risk</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>False</th>\n",
       "      <td>61</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "5yr_efs  high risk  low risk\n",
       "TMG                         \n",
       "False           61        23\n",
       "True            18         1"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a count table\n",
    "count = pd.crosstab(clinical_103['TMG'], clinical_103['5yr_efs'])\n",
    "count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As the count table suggests, there are 61 patients who are non-TMG-mutated who are high-risk, 23 who are non-mutated who are low-risk, 18 who are TMG-mutated with a high risk, and 1 TMG-mutated with a low risk.\n",
    "\n",
    "To verify if the result is statistically significant, we can perform a statistical test, such as Fisher's exact test."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P-value: 0.04010830574733047\n"
     ]
    }
   ],
   "source": [
    "# Perform Fisher's exact test\n",
    "from scipy.stats import fisher_exact\n",
    "\n",
    "oddsr, p = fisher_exact(count, alternative=\"two-sided\")\n",
    "print(\"P-value:\", p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since the p-value is 0.04 (<0.05) we can conclude that the result is statistically significant. Now, we can move on to the second part of the analysis, building ML models and making predictions on NBL risk. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
